### PR TITLE
[BEAM-8951] Stop using nose in load tests

### PR DESCRIPTION
**Pull request by kamilwu**
Beam is moving away from nose. We could change the way of running Python load tests: instead of being subclasses of `unittest.TestCase`, they could be plain Python scripts, just like wordcount examples.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by kamilwu** [comment]
Run Python Load Tests Smoke

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Python Load Tests Smoke

**Comment by kamilwu** [comment]
Run Python Load Tests Smoke

**Comment by kamilwu** [comment]
R: @tvalentyn Can I ask you for a review? 

Without nose, it was much easier to include logging records to console output (https://issues.apache.org/jira/browse/BEAM-8997).

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by tvalentyn** [comment]
Didn't have time to take a look today and I am planning to be out next week. 
@udim have converted several suites to pytest recently and may have some feedback here.

With nose, I think we had to configure output collectors via xml files, see: https://github.com/apache/beam/blob/754b64b4a59f717d84032570acb8ed4cad87b227/sdks/python/scripts/run_integration_test.sh#L248 , I have not yet learned how change output collection with pytest.

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
> With nose, I think we had to configure output collectors via xml files

I think it doesn't really matter in this case since we don't run load tests in parallel.

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

File: sdks/python/apache_beam/testing/load_tests/load_test.py [hunk 1]
```diff
@@ -17,19 +17,50 @@
...
 from apache_beam.metrics import MetricsFilter
 from apache_beam.testing.load_tests.load_test_metrics_utils import MetricsReader
 from apache_beam.testing.test_pipeline import TestPipeline
 
 
-class LoadTest(unittest.TestCase):
-  def parseTestPipelineOptions(self, options=None):
+class LoadTest(object):
+  def __init__(self):
+    self.pipeline = TestPipeline(is_integration_test=True)
```
Comment (tvalentyn) [inline sdks/python/apache_beam/testing/load_tests/load_test.py]: If we proceed with this change, do we still need to use TestPipeline as opposed to regular pipeline?

File: sdks/python/apache_beam/testing/load_tests/streaming/group_by_key_streaming_test.py [hunk 2]
```diff
@@ -62,12 +56,11 @@ def _add_argparse_args(cls, parser):
...
 
 
-@unittest.skipIf(not load_test_enabled, 'Enabled only for phase triggering.')
 class GroupByKeyStreamingTest(LoadTest):
   ID_LABEL = 'id'
 
-  def setUp(self):
-    super(GroupByKeyStreamingTest, self).setUp()
+  def __init__(self):
+    super(GroupByKeyStreamingTest, self).__init__()
```
[the remark this answers is not in the document]
Comment (kamilwu) [reply sdks/python/apache_beam/testing/load_tests/streaming/group_by_key_streaming_test.py]: No, `save_main_session` is not required in any of these tests.

We have one ParDo test running on Python 3 on Dataflow that we can run to make sure everything works with these changes. Could you run it by typing `Run Python 3.7 Load Tests ParDo Dataflow Batch`? I'm afraid that I lost an ability to run Jenkins tests as a non-commiter.

File: sdks/python/apache_beam/testing/load_tests/load_test.py (hunk 1 above)
Comment (kamilwu) [reply sdks/python/apache_beam/testing/load_tests/load_test.py]: It's just a matter of passing pipeline options in a different way (without `--test-pipeline-options` command line argument). Do you think it might be beneficial to use regular pipeline instead of TestPipeline?

**Comment by tvalentyn** [comment]
Run Python 3.7 Load Tests ParDo Dataflow Batch

File: sdks/python/apache_beam/testing/load_tests/load_test.py (hunk 1 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/testing/load_tests/load_test.py]: I'd let @markflyhigh  comment on which way is more idiomatic/preferable going forward for adding new tests. Without test-pipeline-options we will be declaring options explicitly, it may be easier to see which flags need to be passed.

File: sdks/python/apache_beam/testing/load_tests/load_test.py [hunk 3]
```diff
@@ -31,8 +31,8 @@ def __init__(self):
     self.project_id = self.pipeline.get_option('project')
     self.metrics_namespace = self.pipeline.get_option('metrics_table')
     self._metrics_monitor = MetricsReader(
-        publish_to_bq=self.pipeline.get_option('publish_to_big_query') ==
-        'true',
+        publish_to_bq=self.pipeline.get_option('publish_to_big_query')
+        .lower() == 'true',
         project_name=self.project_id,
```
Comment (tvalentyn) [inline sdks/python/apache_beam/testing/load_tests/load_test.py]: My earlier suggestion would also help catch typos, since this flag is not declared a boolean.
If we end up removing test pipeline options, you could also use something like:
```
 parser.add_argument('--publish_to_big_query', default=False,
                      type=lambda s: bool(['false', 'true'].index(s.lower())),
```
It is easy to pass an incorrect value of a boolean flag without noticing the error...

File: sdks/python/apache_beam/testing/load_tests/load_test.py (hunk 1 above)
Comment (tvalentyn) [reply sdks/python/apache_beam/testing/load_tests/load_test.py]: I spoke with @udim - if we migrate these tests to use pytest later, sticking with `--test-pipeline-options` may ease this migration. we haven't fully paved the path to switch to pytest, but @udim may have some in-progress PRs in case you are curious to give pytest a try for loadtest purposes.

Comment (kamilwu) [reply sdks/python/apache_beam/testing/load_tests/load_test.py]: In this case, I'd rather stick with `--test-pipeline-options` for now. Thanks for the link.

File: sdks/python/apache_beam/testing/load_tests/load_test.py (hunk 3 above)
Comment (kamilwu) [reply sdks/python/apache_beam/testing/load_tests/load_test.py]: I suppose I hadn't understood your suggestion fully, my apologies. It's definitely worth implementing. I also added try/except block to make an exception a little more informative for users.

**Comment by kamilwu** [comment]
Rebased the branch with the latest master.

**Comment by kamilwu** [comment]
@tvalentyn Could you trigger the Python 3.7 test once again, please? It should work now.

**Comment by tvalentyn** [comment]
ignore this comment

**Comment by tvalentyn** [comment]
Run Python 3.7 Load Tests ParDo Dataflow Batch

**Comment by Ardagan** [comment]
Run Load Tests Java Combine Portable Flink Batch

**Comment by kkucharc** [comment]
retest this please

**Comment by kkucharc** [comment]
retest this please

**Comment by kkucharc** [comment]
Run Seed Job

**Comment by kkucharc** [comment]
Run Seed Job

**Comment by kkucharc** [comment]
Run Python 3.7 Load Tests ParDo Dataflow Batch

File: sdks/python/apache_beam/testing/load_tests/load_test.py (hunk 3 above)
Comment (markflyhigh) [reply sdks/python/apache_beam/testing/load_tests/load_test.py]: Another way to make those options organized is to register them in `TestPipelineOption` (recommended if commonly used in performance testing), which works the same way as parser does.

File: sdks/python/apache_beam/testing/load_tests/load_test.py [hunk 4]
```diff
@@ -18,18 +18,65 @@
...
 
 from apache_beam.metrics import MetricsFilter
 from apache_beam.testing.load_tests.load_test_metrics_utils import MetricsReader
 from apache_beam.testing.test_pipeline import TestPipeline
 
 
-class LoadTest(unittest.TestCase):
-  def parseTestPipelineOptions(self, options=None):
+class LoadTest(object):
+  def __init__(self):
```
[the remark this answers is not in the document]
Comment (kamilwu) [reply sdks/python/apache_beam/testing/load_tests/load_test.py]: No particular benefit. Because `LoadTest` is no longer a subclass of `unittest.TestCase`, I had to move this code somewhere. And the reasons why I stopped using `unittest.TestCase` are:
- We had to prevent load tests from being included in other test suites and that's why `LOAD_TEST_ENABLED` guard was used. Now the guard is no longer needed,
- `tearDown()` was actually used to run a pipeline, which was misleading, as tearDown is supposed only to clean up resources after test execution,
- Without nose or pytest, it is difficult to pass extra command-line arguments.

To sum up, I think `unittest` doesn't suit our needs here and can be easily replaced with plain Python classes.

File: sdks/python/apache_beam/testing/load_tests/load_test.py (hunk 3 above)
Comment (kamilwu) [reply sdks/python/apache_beam/testing/load_tests/load_test.py]: I created `LoadTestOptions` class with common parameters that load tests can take. `--publish_to_big_query` is one of them. I hope this will make those options better organised.

**Comment by kkucharc** [comment]
retest this please

**Comment by kkucharc** [comment]
run seed job

**Comment by kkucharc** [comment]
Run Python 3.7 Load Tests ParDo Dataflow Batch

**Comment by kkucharc** [comment]
retest this please

**Comment by kkucharc** [comment]
run seed job

**Comment by kkucharc** [comment]
run seed job

**Comment by kkucharc** [comment]
Run Python 3.7 Load Tests ParDo Dataflow Batch

**Comment by tvalentyn** [comment]
@kamilwu, once this is ready for the next step review iteration please say PTAL @reviewer_name(,s) or something to make sure reviewers are aware that you are waiting on their input. Thank you.

**Comment by kkucharc** [comment]
retest this please

**Comment by kamilwu** [comment]
@tvalentyn Thanks, I'm aware of this. Right now I'm struggling to make all checks green again

**Comment by kamilwu** [comment]
Merge conflicts resolved.

**Comment by kkucharc** [comment]
Run Python PreCommit

**Comment by kkucharc** [comment]
Run Seed Job

**Comment by kkucharc** [comment]
Run Python 3.7 Load Tests ParDo Dataflow Batch

**Comment by kkucharc** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
@tvalentyn @markflyhigh PTAL

**Comment by kamilwu** [comment]
R: @tvalentyn @markflyhigh Kindly asking for a review once again

**Comment by tvalentyn** [comment]
Sorry for the delay and if I wasn't explicit enough - I'd like @markflyhigh to take a first pass on these changes. I'll touch bases with Mark to see if he is available to review.

**Comment by markflyhigh** [comment]
Hi Kamil, sorry for the delay. I'm occupied by oncall stuff this week and plan to get back to this PR on Tuesday next week. 

I understand the idea of removing nose from load test. However, this change seems more than that, which is moving load test out of python test, and creating a divergence between general integration test. Can you provide more information for why we want to do that? What's the benefit? How does it affect the test management? Thanks!

**Comment by kamilwu** [comment]
Thanks @markflyhigh.

I've already pointed out the benefits in one of the comments above, let me do this once again here.

What do we gain from moving load tests out of python test:
-  Without nose, it is difficult to pass additional command-line arguments to the test. Since load tests require additional parameters, this looks like a blocker for us.
- As long as load tests are based on python tests, they are discovered by nose/pytest and included in test suites running as a part of PreCommit. This is undesirable. Load tests require additional parameters and they were never intented to be run as a part of PreCommit. In order to prevent this behaviour, we had developed some kind of guard[1] which skips test if certain condition is met. If we move out of python test, this guard will not be needed anymore.

In general, this does not affect test management that much. We use gradle task[2] to run load tests and this gradle task is the only place affected by my changes.

Let me know if you find this answer helpful.

[1] https://github.com/apache/beam/blob/31a5a33bd268eb93aa5eb8d1c31f5043b85b1511/sdks/python/apache_beam/testing/load_tests/pardo_test.py#L134-L139
[2] https://github.com/apache/beam/blob/master/sdks/python/apache_beam/testing/load_tests/build.gradle

**Comment by mwalenia** [comment]
retest this poelase

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run Python 3.7 Load Tests ParDo Dataflow Batch

**Comment by kamilwu** [comment]
R: @markflyhigh Please take a look at my answer. It'd be good if we could move forward with the PR

**Comment by kamilwu** [comment]
@markflyhigh Thanks for this in-depth explanation. Right now I'm distracted by other stuff too, but I'm really interested in helping to unify the way we construct those tests in Python SDK. Keep me posted about the progress, if there would be any.
